### ISSUE TITLE
CORS issue in login flow

### ISSUE THREAD

**Issue by salonikumawat28** [nextcloud/server#34898]
*Context*
I have a website(not the NextCloud app) from which I want to login into NextCloud so that I can get credentials which I can use to download and upload files.

First step I did was using the hardcoded username:password and passed it as `Authorization` header to download and upload URLs which failed on CORS. I was able to resolve the CORS issues once I installed `WebAppPassword` and whitelisted my website.

As a second step, I tried the [login v2 flow](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2) and it failed on CORS. This was surprising as my website origin was already whitelisted in`WebAppPassword` and was working for download and upload URLs.

As a temporary hack, I added `@CORS` annotation in the login v2 controllers and it worked on all URLs except `login/v2/poll` because it fails on `preflight` request as it's a `POST` call with json content type. To solve this, another temporary hack I did is to add preflight URL for login v2 flow.

*Problem*
As these hacks were direct code changes in the NextCloud server code, I want to replace these hacks with a proper solutions. Things I tried:
1. I created an app within the NextCloud and registered a Middleware to resolve the CORS issues. The problem is that the Middleware works only the URLs pointed to the app I wrote and it doesn't work on the login URLs. Even if it works somehow, it will not resolve the preflight issue.

Can you please help me on how to resolve the CORS and preflight issues?

**Comment by salonikumawat28** [comment]
I looked into similar issues like https://github.com/nextcloud/server/issues/3131 and all the solution attempts and temporary hacks mentioned in the issue were based on the suggested mentioned in the issue, NextCloud [login v2 flow](https://docs.nextcloud.com/server/stable/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2) doc and NextCloud [OAuth flow](https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/oauth2.html) doc but I was not able to find a solution which doesn't involve changes in NextCloud server.

**Comment by salonikumawat28** [comment]
Gentle Reminde : Can someone help me into this ?

**Comment by salonikumawat28** [comment]
Can please someone help me on this issue?

**Comment by julien-nc** [comment]
@juliushaertl Not really, no. The only topic where it is currently possible to canonically deal with CORS headers is WebDAV endpoints (by adding a Sabre plugin like the WebAppPassword app does).

@salonikumawat28 By the way, out of topic, if you want to avoid having WebAppPassword as a dependency, you can easily implement a similar Sabre plugin in your app that will whitelist your website's domain.

IMO you can avoid dealing with CORS issues if you implement the login flow v2 or OAuth2 code flow using your website's backend which will be able to reach the "/token" endpoint. Once you get an OAuth token (or an app password), you can provide it to your frontend which can then use it for requests to WebDAV endpoints (CORS being dealt with by WebAppPassword).

But maybe you have a strong constraint to perform the auth flows entirely in the browser.

The integration_* apps implement the OAuth2 code flow (to connect to external services) using the NC backend to perform the last `/token` request.

So for you, the OAuth implementation would be:
* your website's frontend opens a new tab (or redirects) to the authorize endpoint
* after the users have authenticated in NC, they get redirected to the redirection URL which is handled by your backend
* your backend does the last `/apps/oauth2/api/v1/token` request to get the token
* the token is provided to your frontend which can then perform WebDAV requests

**Comment by beatles1** [comment]
I'm also confused by this, I've made a browser extension for someone else's Nextcloud app. I want to be able to use the login v2 flow but because my app is in a browser and respects CORS I  can't without asking the user for full permission to the servers domain. That means that the session in my extension and the session in the browser are one and the same and so I've not got my own session.

I don't understand why the login flow would be configured to only work for the same origin, I'd assume using it from other origins is one of the main reasons it exists. Of course this works fine from a mobile app or something which doesn't respect CORS but just not for client only webapps.

**Comment by Chalolennox** [comment]
Is someone still working on this? Honestly, this is an essential feature.

**Comment by Lichtblitz** [comment]
Nexcloud looked very promising on the surface. Now that we're nearing production, we notice more and more slightly broken features with years-old open tickets. Nexcloud has a login flow that requires same-origin. It's the standard login mask. The login v2 should be called from external applications. That's the (more or less) only reason for it to exist. Right now it is in an unusable state if the dev is not circumventing security measures or using frameworks that ignore them in the first place.
